<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.340">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>HoloFood Data Portal Documentation - API for Programmatic Access</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./database.html" rel="next">
<link href="./website.html" rel="prev">
<link href="./img/hf-logo-brand.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script>
<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">HoloFood Data Portal Documentation</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://www.holofooddata.org" rel="" target="">
 <span class="menu-text">Data Portal</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./api.html">API for Programmatic Access</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
        <div class="sidebar-tools-collapse">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-collapse-item" href="https://github.com/ebi-metagenomics/holofood-database">
            Data Portal Repository
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-collapse-item" href="https://github.com/ebi-metagenomics/holofood-database/docs">
            Documentation source
            </a>
          </li>
      </ul>
    </div>
    <a href="mailto:holofood-help@ebi.ac.uk" title="Help" class="quarto-navigation-tool px-1" aria-label="Help"><i class="bi bi-question-circle-fill"></i></a>
</div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link"><i class="bi bi-box" role="img">
</i> 
 <span class="menu-text">Docs home</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Datasets</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./website.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Browsing the website</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./api.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">API for Programmatic Access</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Database snapshots</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tutorial</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./publications.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Publications</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./partners.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Admin access for HoloFood partners</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#canonical-urls" id="toc-canonical-urls" class="nav-link active" data-scroll-target="#canonical-urls">Canonical URLs</a></li>
  <li><a href="#api-endpoints-and-playground" id="toc-api-endpoints-and-playground" class="nav-link" data-scroll-target="#api-endpoints-and-playground">API Endpoints and Playground</a>
  <ul class="collapse">
  <li><a href="#apisamples" id="toc-apisamples" class="nav-link" data-scroll-target="#apisamples"><code>/api/samples</code></a></li>
  <li><a href="#apianimals" id="toc-apianimals" class="nav-link" data-scroll-target="#apianimals"><code>/api/animals</code></a></li>
  <li><a href="#apianalysis-summaries" id="toc-apianalysis-summaries" class="nav-link" data-scroll-target="#apianalysis-summaries"><code>/api/analysis-summaries</code></a></li>
  <li><a href="#apigenome-catalogues" id="toc-apigenome-catalogues" class="nav-link" data-scroll-target="#apigenome-catalogues"><code>/api/genome-catalogues</code></a></li>
  <li><a href="#apiviral-catalogues" id="toc-apiviral-catalogues" class="nav-link" data-scroll-target="#apiviral-catalogues"><code>/api/viral-catalogues</code></a></li>
  </ul></li>
  <li><a href="#using-the-api" id="toc-using-the-api" class="nav-link" data-scroll-target="#using-the-api">Using the API</a>
  <ul class="collapse">
  <li><a href="#from-the-command-line" id="toc-from-the-command-line" class="nav-link" data-scroll-target="#from-the-command-line">From the command line</a></li>
  <li><a href="#from-python" id="toc-from-python" class="nav-link" data-scroll-target="#from-python">From Python</a></li>
  </ul></li>
  <li><a href="#filters-using-query-parameters" id="toc-filters-using-query-parameters" class="nav-link" data-scroll-target="#filters-using-query-parameters">Filters using query parameters</a></li>
  <li><a href="#mgnify" id="toc-mgnify" class="nav-link" data-scroll-target="#mgnify">MGnify</a></li>
  <li><a href="#metabolights" id="toc-metabolights" class="nav-link" data-scroll-target="#metabolights">MetaboLights</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.dev/ebi-metagenomics/holofood-database/blob/main/docs/api.ipynb" class="toc-action">Edit this page</a></p><p><a href="https://github.com/ebi-metagenomics/holofood-database/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">API for Programmatic Access</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p>APIs (Application Programming Interfaces) allow programmatic access to data resources. If you need to fetch and analyse metadata for many Samples, you can write a script to fetch data in this way.</p>
<p>The <a href="">HoloFood Data Portal API</a> gives programmatic access to the HoloFood Samples and their metadata, as well as URLs for where datasets are stored in public archives.</p>
<section id="canonical-urls" class="level2">
<h2 class="anchored" data-anchor-id="canonical-urls">Canonical URLs</h2>
<p>Throughout the API, <code>canonical_url</code>s are returned which point to the canonical database entry, i.e.&nbsp;the authoritative source, for each data object.</p>
<p>These are:</p>
<ul>
<li><a href="https://www.ebi.ac.uk/biosamples">BioSamples</a> for Samples (both host-animal level and derived-extraction level).</li>
<li>The <a href="https://www.ebi.ac.uk/ena/browser/home">European Nucleotide Archive Browser</a> for nucleotide sequencing data.</li>
<li><a href="https://www.ebi.ac.uk/metagenomics">MGnify</a> for metagenomic-derived analyses and MAGs (metagenome assembled genomes)</li>
<li><a href="https://www.ebi.ac.uk/metabolights/">MetaboLights</a> for metabolomic analyses</li>
<li>The websites of various partner institutions and registries where an <a href="https://en.wikipedia.org/wiki/Internationalized_Resource_Identifier">IRI</a> has been supplied with a metadata entry.</li>
<li>The HoloFood Data Portal itself for “Analysis Summaries”, which are documents hosted only by this website.</li>
<li>The HoloFood Data Portal itself for viral annotations</li>
</ul>
</section>
<section id="api-endpoints-and-playground" class="level2">
<h2 class="anchored" data-anchor-id="api-endpoints-and-playground">API Endpoints and Playground</h2>
<p>The <a href="https://www.holofooddata.org/api">“Browsable API Playground”</a> is the best place to discover the API. Find this under the <code>API</code> navigation item on the data portal.</p>
<p>The browsable API lets you see the endpoints and their response schemas. Under an endpoint, press <code>Try it out</code> to see the output for a specific query.</p>
<p>In brief, the top-level endpoints are:</p>
<section id="apisamples" class="level3">
<h3 class="anchored" data-anchor-id="apisamples"><code>/api/samples</code></h3>
<p>List samples, or fetch details about a specific sample (like its metadata). List all possible metadata markers (i.e.&nbsp;keys).</p>
</section>
<section id="apianimals" class="level3">
<h3 class="anchored" data-anchor-id="apianimals"><code>/api/animals</code></h3>
<p>List animals (hosts), or fetch details about a specific animal (like its metadata and derived samples).</p>
</section>
<section id="apianalysis-summaries" class="level3">
<h3 class="anchored" data-anchor-id="apianalysis-summaries"><code>/api/analysis-summaries</code></h3>
<p>List summary analyses published on the data portal.</p>
</section>
<section id="apigenome-catalogues" class="level3">
<h3 class="anchored" data-anchor-id="apigenome-catalogues"><code>/api/genome-catalogues</code></h3>
<p>List MAG catalogues, or fetch detail about a catalogue, or list the MAGs within a catalogue.</p>
</section>
<section id="apiviral-catalogues" class="level3">
<h3 class="anchored" data-anchor-id="apiviral-catalogues"><code>/api/viral-catalogues</code></h3>
<p>List Viral catalogues, or fetch detail about a catalogue, or list the fragments within a catalogue.</p>
</section>
</section>
<section id="using-the-api" class="level2">
<h2 class="anchored" data-anchor-id="using-the-api">Using the API</h2>
<section id="from-the-command-line" class="level3">
<h3 class="anchored" data-anchor-id="from-the-command-line">From the command line</h3>
<p>Use a command line tool like <a href="https://curl.se/">cURL</a> to query the API. Responses are in <a href="https://www.json.org/json-en.html">JSON</a> format.</p>
<p>For example to list all samples:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> https://www.holofooddata.org/api/samples</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>To find the <code>Stearic acid 18:0</code> data associated with Fatty Acids sample <code>SAMEA112949944</code>, we could (using <a href="https://stedolan.github.io/jq/"><code>jq</code></a> to handle JSON data on the command line):</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">curl</span> https://www.holofooddata.org/api/samples/SAMEA112949944 <span class="kw">|</span> <span class="ex">jq</span> <span class="st">'.structured_metadata | .[] | select(.marker.name == "Stearic acid 18:0")'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>and get</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode json code-with-copy"><code class="sourceCode json"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">{</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="dt">"marker"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"name"</span><span class="fu">:</span> <span class="st">"Stearic acid 18:0"</span><span class="fu">,</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"type"</span><span class="fu">:</span> <span class="st">"FATTY ACIDS MG"</span><span class="fu">,</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"canonical_url"</span><span class="fu">:</span> <span class="kw">null</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">},</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>  <span class="dt">"measurement"</span><span class="fu">:</span> <span class="st">"1,47"</span><span class="fu">,</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  <span class="dt">"units"</span><span class="fu">:</span> <span class="st">"mg/g"</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="fu">}</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="fu">{</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>  <span class="dt">"marker"</span><span class="fu">:</span> <span class="fu">{</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"name"</span><span class="fu">:</span> <span class="st">"Stearic acid 18:0"</span><span class="fu">,</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"type"</span><span class="fu">:</span> <span class="st">"FATTY ACIDS PERCENTAGE"</span><span class="fu">,</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    <span class="dt">"canonical_url"</span><span class="fu">:</span> <span class="kw">null</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">},</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>  <span class="dt">"measurement"</span><span class="fu">:</span> <span class="st">"1,47"</span><span class="fu">,</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>  <span class="dt">"units"</span><span class="fu">:</span> <span class="st">"%"</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="fu">}</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="from-python" class="level3">
<h3 class="anchored" data-anchor-id="from-python">From Python</h3>
<p>More realistically, to fetch a list of samples as a <a href="https://pandas.pydata.org/">Pandas dataframe</a></p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Packages required
</div>
</div>
<div class="callout-body-container callout-body">
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install requests pandas</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> requests</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>samples <span class="op">=</span> requests.get(<span class="st">'https://www.holofooddata.org/api/samples'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>samples_df <span class="op">=</span> pd.json_normalize(samples.json()[<span class="st">'items'</span>])</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>samples_df.head(<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">accession</th>
<th data-quarto-table-cell-role="th">title</th>
<th data-quarto-table-cell-role="th">sample_type</th>
<th data-quarto-table-cell-role="th">animal</th>
<th data-quarto-table-cell-role="th">canonical_url</th>
<th data-quarto-table-cell-role="th">metagenomics_url</th>
<th data-quarto-table-cell-role="th">metabolomics_url</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>SAMEA10104908</td>
<td>CA01.07F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112905066</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>SAMEA10104910</td>
<td>CA02.12F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112904813</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>SAMEA10104911</td>
<td>CA02.18F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112904777</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div class="callout callout-style-default callout-important callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Paginated data
</div>
</div>
<div class="callout-body-container callout-body">
<p>However… we haven’t got all of the data. We only have one page.</p>
<p>The <code>?page=</code> URL query parameter lets us retrieve subsequent pages.</p>
</div>
</div>
<div class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(samples_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>100</code></pre>
</div>
</div>
<p>The API response does tell us how many items there are <em>in total</em>:</p>
<div class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>samples.json()[<span class="st">'count'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
<pre><code>9889</code></pre>
</div>
</div>
<div class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>samples_endpoint_base <span class="op">=</span> <span class="st">'https://www.holofooddata.org/api/samples'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>page <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># We will only fetch the first 10 pages for now...</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>max_pages <span class="op">=</span> <span class="dv">10</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="cf">while</span> page <span class="kw">and</span> page <span class="op">&lt;</span> max_pages:</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Fetching </span><span class="sc">{</span>page<span class="op">=</span><span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>    samples_page <span class="op">=</span> requests.get(<span class="ss">f'</span><span class="sc">{</span>samples_endpoint_base<span class="sc">}</span><span class="ss">?</span><span class="sc">{</span>page<span class="op">=</span><span class="sc">}</span><span class="ss">'</span>).json()</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>    samples_page_df <span class="op">=</span> pd.json_normalize(samples_page[<span class="st">'items'</span>])</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> page <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>        samples_df <span class="op">=</span> samples_page_df</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>        samples_df <span class="op">=</span> pd.concat([samples_df, samples_page_df])</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>    page <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(samples_df) <span class="op">&gt;=</span> samples_page[<span class="st">'count'</span>]:</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>        page <span class="op">=</span> <span class="va">False</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Fetching page=1
Fetching page=2
Fetching page=3
Fetching page=4
Fetching page=5
Fetching page=6
Fetching page=7
Fetching page=8
Fetching page=9</code></pre>
</div>
</div>
<div class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(samples_df)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="10">
<pre><code>900</code></pre>
</div>
</div>
<section id="find-all-metagenomic-assembly-samples" class="level4">
<h4 class="anchored" data-anchor-id="find-all-metagenomic-assembly-samples">Find all <em>metagenomic assembly</em> samples:</h4>
<div class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>samples_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">accession</th>
<th data-quarto-table-cell-role="th">title</th>
<th data-quarto-table-cell-role="th">sample_type</th>
<th data-quarto-table-cell-role="th">animal</th>
<th data-quarto-table-cell-role="th">canonical_url</th>
<th data-quarto-table-cell-role="th">metagenomics_url</th>
<th data-quarto-table-cell-role="th">metabolomics_url</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>SAMEA10104908</td>
<td>CA01.07F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112905066</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>SAMEA10104910</td>
<td>CA02.12F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112904813</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>SAMEA10104911</td>
<td>CA02.18F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112904777</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>SAMEA10104912</td>
<td>CA03.10F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112904752</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>SAMEA10104914</td>
<td>CA04.10F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112904915</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">95</td>
<td>SAMEA112369857</td>
<td>SB11.04B1a</td>
<td>transcriptomic</td>
<td>SAMEA112949107</td>
<td>https://www.ebi.ac.uk/biosamples/SAMEA112369857</td>
<td>None</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">96</td>
<td>SAMEA112369858</td>
<td>SB11.05B1a</td>
<td>transcriptomic</td>
<td>SAMEA112949088</td>
<td>https://www.ebi.ac.uk/biosamples/SAMEA112369858</td>
<td>None</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">97</td>
<td>SAMEA112369859</td>
<td>SB11.11B1a</td>
<td>transcriptomic</td>
<td>SAMEA112949878</td>
<td>https://www.ebi.ac.uk/biosamples/SAMEA112369859</td>
<td>None</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">98</td>
<td>SAMEA112369860</td>
<td>SB11.12B1a</td>
<td>transcriptomic</td>
<td>SAMEA112948858</td>
<td>https://www.ebi.ac.uk/biosamples/SAMEA112369860</td>
<td>None</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">99</td>
<td>SAMEA112369861</td>
<td>SB11.13B1a</td>
<td>transcriptomic</td>
<td>SAMEA112948701</td>
<td>https://www.ebi.ac.uk/biosamples/SAMEA112369861</td>
<td>None</td>
<td>None</td>
</tr>
</tbody>
</table>

<p>900 rows × 7 columns</p>
</div>
</div>
</div>
<div class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>metagenome_assembly_samples <span class="op">=</span> samples_df[samples_df.sample_type <span class="op">==</span> <span class="st">'metagenomic_assembly'</span>]</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="bu">len</span>(metagenome_assembly_samples))</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>metagenome_assembly_samples.head(<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>458</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="16">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">accession</th>
<th data-quarto-table-cell-role="th">title</th>
<th data-quarto-table-cell-role="th">sample_type</th>
<th data-quarto-table-cell-role="th">animal</th>
<th data-quarto-table-cell-role="th">canonical_url</th>
<th data-quarto-table-cell-role="th">metagenomics_url</th>
<th data-quarto-table-cell-role="th">metabolomics_url</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>SAMEA10104908</td>
<td>CA01.07F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112905066</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>SAMEA10104910</td>
<td>CA02.12F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112904813</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>SAMEA10104911</td>
<td>CA02.18F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112904777</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
</section>
</section>
<section id="filters-using-query-parameters" class="level2">
<h2 class="anchored" data-anchor-id="filters-using-query-parameters">Filters using query parameters</h2>
<p>We didn’t need to fetch all the samples to then select only the metagenomic ones. We can instead fetch only the metagenomic ones using a <strong>query parameter filter</strong>. (If you’re familiar with SQL, this eventually maps to a <code>WHERE</code> clause on the database.)</p>
<div class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>page <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="cf">while</span> page:</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'Fetching </span><span class="sc">{</span>page<span class="op">=</span><span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>    mg_samples_page <span class="op">=</span> requests.get(<span class="ss">f'</span><span class="sc">{</span>samples_endpoint_base<span class="sc">}</span><span class="ss">?</span><span class="sc">{</span>page<span class="op">=</span><span class="sc">}</span><span class="ss">&amp;sample_type=metagenomic_assembly'</span>).json()</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>    mg_samples_page_df <span class="op">=</span> pd.json_normalize(mg_samples_page[<span class="st">'items'</span>])</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> page <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>        mg_samples_df <span class="op">=</span> mg_samples_page_df</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>        mg_samples_df <span class="op">=</span> pd.concat([mg_samples_df, mg_samples_page_df])</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>    page <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">len</span>(mg_samples_df) <span class="op">&gt;=</span> mg_samples_page[<span class="st">'count'</span>]:</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>        page <span class="op">=</span> <span class="va">False</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Fetching page=1
Fetching page=2
Fetching page=3
Fetching page=4
Fetching page=5
Fetching page=6
Fetching page=7
Fetching page=8
Fetching page=9
Fetching page=10
Fetching page=11
Fetching page=12
Fetching page=13
Fetching page=14</code></pre>
</div>
</div>
<div class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>samples_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="18">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">accession</th>
<th data-quarto-table-cell-role="th">title</th>
<th data-quarto-table-cell-role="th">sample_type</th>
<th data-quarto-table-cell-role="th">animal</th>
<th data-quarto-table-cell-role="th">canonical_url</th>
<th data-quarto-table-cell-role="th">metagenomics_url</th>
<th data-quarto-table-cell-role="th">metabolomics_url</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>SAMEA10104908</td>
<td>CA01.07F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112905066</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>SAMEA10104910</td>
<td>CA02.12F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112904813</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>SAMEA10104911</td>
<td>CA02.18F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112904777</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>SAMEA10104912</td>
<td>CA03.10F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112904752</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>SAMEA10104914</td>
<td>CA04.10F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112904915</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA10...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">78</td>
<td>SAMEA9449961</td>
<td>CC17.06F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112905387</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA94...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">79</td>
<td>SAMEA9449962</td>
<td>CC17.10F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112905371</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA94...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">80</td>
<td>SAMEA9449963</td>
<td>CC16.16F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112905164</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA94...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">81</td>
<td>SAMEA9449964</td>
<td>CC18.03F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112904811</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA94...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">82</td>
<td>SAMEA9449965</td>
<td>CC18.05F1a</td>
<td>metagenomic_assembly</td>
<td>SAMEA112905140</td>
<td>https://www.ebi.ac.uk/ena/browser/view/SAMEA94...</td>
<td>https://www.ebi.ac.uk/metagenomics/api/v1/samp...</td>
<td>None</td>
</tr>
</tbody>
</table>

<p>1383 rows × 7 columns</p>
</div>
</div>
</div>
<p>And so, we’ve more efficiently fetch the metagenomic assemblies – by not pulling pages and pages of irrelevant data.</p>
</section>
<section id="mgnify" class="level1">
<h1>MGnify</h1>
<p>You can obtain information about the analyses present on MGnify for a certain sample, using the <code>metagenomic_url</code> returned by the data portal API:</p>
<div class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>response <span class="op">=</span> requests.get(<span class="st">'https://www.holofooddata.org/api/samples/SAMEA13604493'</span>)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>mgnify_url <span class="op">=</span> response.json().get(<span class="st">'metagenomics_url'</span>)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>mgnify_url</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="27">
<pre><code>'https://www.ebi.ac.uk/metagenomics/api/v1/samples/SAMEA13604493'</code></pre>
</div>
</div>
<p>You can then use that <a href="https://www.ebi.ac.uk/metagenomics/api/v1">MGnify API</a> endopint to fetch information about that sample and follow related links through MGnify:</p>
<div class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>mgnify_response <span class="op">=</span> requests.get(mgnify_url)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>mgnify_data <span class="op">=</span> mgnify_response.json().get(<span class="st">'data'</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>runs_url <span class="op">=</span> mgnify_data[<span class="st">'relationships'</span>][<span class="st">'runs'</span>][<span class="st">'links'</span>][<span class="st">'related'</span>]</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>mgnify_runs_response <span class="op">=</span> requests.get(runs_url)</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>mgnify_runs_data <span class="op">=</span> mgnify_runs_response.json().get(<span class="st">'data'</span>)</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>mgnify_runs_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="33">
<pre><code>[{'type': 'runs',
  'id': 'ERR9358532',
  'attributes': {'experiment-type': 'metagenomic',
   'is-private': False,
   'accession': 'ERR9358532',
   'secondary-accession': 'ERR9358532',
   'ena-study-accession': None,
   'instrument-platform': 'ILLUMINA',
   'instrument-model': 'Illumina NovaSeq 6000'},
  'relationships': {'analyses': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/runs/ERR9358532/analyses'}},
   'assemblies': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/runs/ERR9358532/assemblies'},
    'data': [{'type': 'assemblies',
      'id': 'ERZ13633665',
      'links': {'self': 'https://www.ebi.ac.uk/metagenomics/api/v1/assemblies/ERZ13633665'}}]},
   'study': {'data': {'type': 'studies', 'id': 'MGYS00006086'},
    'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00006086'}},
   'sample': {'data': {'type': 'samples', 'id': 'ERS11206669'},
    'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/samples/ERS11206669'}},
   'pipelines': {'data': []}},
  'links': {'self': 'https://www.ebi.ac.uk/metagenomics/api/v1/runs/ERR9358532'}}]</code></pre>
</div>
</div>
<hr>
<p>In most cases, you can also use the base endpoint you’re interested in (e.g.&nbsp;<code>/analyses</code>) with a filter based on the sample accession. E.g. to find a MGnify analyses of a sample:</p>
<div class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>mgnify_response <span class="op">=</span> requests.get(<span class="st">'https://www.ebi.ac.uk/metagenomics/api/v1/analyses?sample_accession=SAMEA13604493'</span>)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>mgnify_response.json().get(<span class="st">'data'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="36">
<pre><code>[{'type': 'analysis-jobs',
  'id': 'MGYA00617383',
  'attributes': {'analysis-status': 'completed',
   'accession': 'MGYA00617383',
   'experiment-type': 'assembly',
   'analysis-summary': [{'key': 'Submitted nucleotide sequences',
     'value': '344'},
    {'key': 'Nucleotide sequences after format-specific filtering',
     'value': '344'},
    {'key': 'Nucleotide sequences after length filtering', 'value': '344'},
    {'key': 'Nucleotide sequences after undetermined bases filtering',
     'value': '344'},
    {'key': 'Reads with predicted CDS', 'value': '330'},
    {'key': 'Reads with predicted RNA', 'value': '13'},
    {'key': 'Reads with InterProScan match', 'value': '108'},
    {'key': 'Predicted CDS', 'value': '531'},
    {'key': 'Predicted CDS with InterProScan match', 'value': '127'},
    {'key': 'Total InterProScan matches', 'value': '299'},
    {'key': 'Predicted SSU sequences', 'value': '0'},
    {'key': 'Predicted LSU sequences', 'value': '2'}],
   'pipeline-version': '5.0',
   'is-private': False,
   'complete-time': '2022-12-01T19:35:07',
   'instrument-platform': 'ILLUMINA',
   'instrument-model': 'Illumina NovaSeq 6000'},
  'relationships': {'taxonomy-lsu': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00617383/taxonomy/lsu'}},
   'go-slim': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00617383/go-slim'}},
   'antismash-gene-clusters': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00617383/antismash-gene-clusters'}},
   'genome-properties': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00617383/genome-properties'}},
   'run': {'data': None},
   'study': {'data': {'type': 'studies', 'id': 'MGYS00006086'},
    'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/studies/MGYS00006086'}},
   'go-terms': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00617383/go-terms'}},
   'taxonomy-itsunite': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00617383/taxonomy/unite'}},
   'sample': {'data': {'type': 'samples', 'id': 'ERS11206669'},
    'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/samples/ERS11206669'}},
   'taxonomy-itsonedb': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00617383/taxonomy/itsonedb'}},
   'taxonomy': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00617383/taxonomy'}},
   'assembly': {'data': {'type': 'assemblies', 'id': 'ERZ13633665'},
    'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/assemblies/ERZ13633665'}},
   'downloads': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00617383/downloads'}},
   'interpro-identifiers': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00617383/interpro-identifiers'}},
   'taxonomy-ssu': {'links': {'related': 'https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00617383/taxonomy/ssu'}}},
  'links': {'self': 'https://www.ebi.ac.uk/metagenomics/api/v1/analyses/MGYA00617383'}}]</code></pre>
</div>
</div>
</section>
<section id="metabolights" class="level1">
<h1>MetaboLights</h1>
<p>You can find the MetaboLights URL for a study from the portal API:</p>
<div class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>mtbls_response <span class="op">=</span> requests.get(<span class="st">'https://www.holofooddata.org/api/samples/SAMEA112907471'</span>)</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>mtbls_url <span class="op">=</span> mtbls_response.json().get(<span class="st">'metabolomics_url'</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>mtbls_url</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="39">
<pre><code>'https://www.ebi.ac.uk/metabolights/ws/studies/MTBLS6988'</code></pre>
</div>
</div>
<p>This URL is for the entire study on MetaboLights, rather than a specific sample. The data portal <strong>website</strong> (but not the API, for performance reasons) navigate the MetaboLights API in order to find the files relevant to a particular sample.</p>
<p>The <a href="https://github.com/EBI-Metagenomics/holofood-database/blob/main/holofood/external_apis/metabolights/api.py">code that does this in the data portal</a> shows how this can be achieved. In short: fetch the files list from the MetaboLights API, and find the <code>metadata_sample</code> file (it is an ISA-TAB file). Look through that file for the the row where <code>Characteristics[BioSamples accession]</code> is your sample(s) of interest. This field may have slightly different names in different studies. This lets you find the <code>Sample Name</code> corresponding to the BioSample accession.</p>
<p>Then, find the <code>metadata_assay</code> file(s) from the file list. Concatenate all of these <code>metadata_assay</code> tab files, and find all the rows from all of the <code>metadata_assay</code> sheets where the <code>Sample Name</code> matches.</p>
<p>These assay sheets will have columns of interest about the assays used, the raw and derived files, and the metabolites association files. These are filenames which you can then download.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./website.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Browsing the website</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./database.html" class="pagination-link">
        <span class="nav-page-text">Database snapshots</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">This project has received funding from the European Union’s Horizon 2020 research and innovation programme under grant agreement No 817729</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>