<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.361">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>HoloFood Data Portal Documentation - Browsing the website</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./api.html" rel="next">
<link href="./datasets.html" rel="prev">
<link href="./img/hf-logo-brand.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">HoloFood Data Portal Documentation</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://www.holofooddata.org" rel="" target="">
 <span class="menu-text">Data Portal</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./website.html">Browsing the website</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
        <div class="sidebar-tools-collapse">
    <div class="dropdown">
      <a href="" title="" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label=""><i class="bi bi-github"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-collapse-item" href="https://github.com/ebi-metagenomics/holofood-database">
            Data Portal Repository
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-collapse-item" href="https://github.com/ebi-metagenomics/holofood-database/docs">
            Documentation source
            </a>
          </li>
      </ul>
    </div>
    <a href="mailto:holofood-help@ebi.ac.uk" title="Help" class="quarto-navigation-tool px-1" aria-label="Help"><i class="bi bi-question-circle-fill"></i></a>
</div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link"><i class="bi bi-box" role="img">
</i> 
 <span class="menu-text">Docs home</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./datasets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Datasets</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./website.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Browsing the website</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./api.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">API for Programmatic Access</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./database.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Database snapshots</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tutorial.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Tutorial</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./publications.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Publications</span></a>
  </div>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./partners.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Admin access for HoloFood partners</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#global-search" id="toc-global-search" class="nav-link active" data-scroll-target="#global-search">Global Search</a></li>
  <li><a href="#finding-samples" id="toc-finding-samples" class="nav-link" data-scroll-target="#finding-samples">Finding Samples</a>
  <ul class="collapse">
  <li><a href="#treatment-search" id="toc-treatment-search" class="nav-link" data-scroll-target="#treatment-search">Treatment search</a></li>
  <li><a href="#sample-detail" id="toc-sample-detail" class="nav-link" data-scroll-target="#sample-detail">Sample detail</a></li>
  </ul></li>
  <li><a href="#downloading-sample-lists-and-metadata" id="toc-downloading-sample-lists-and-metadata" class="nav-link" data-scroll-target="#downloading-sample-lists-and-metadata">Downloading sample lists and metadata</a></li>
  <li><a href="#finding-sample-data-in-other-public-repositories" id="toc-finding-sample-data-in-other-public-repositories" class="nav-link" data-scroll-target="#finding-sample-data-in-other-public-repositories">Finding sample data in other public repositories</a>
  <ul class="collapse">
  <li><a href="#metagenomics" id="toc-metagenomics" class="nav-link" data-scroll-target="#metagenomics">Metagenomics</a></li>
  <li><a href="#metabolomics" id="toc-metabolomics" class="nav-link" data-scroll-target="#metabolomics">Metabolomics</a></li>
  </ul></li>
  <li><a href="#finding-samples-related-to-the-same-host-animal" id="toc-finding-samples-related-to-the-same-host-animal" class="nav-link" data-scroll-target="#finding-samples-related-to-the-same-host-animal">Finding samples related to the same host animal</a></li>
  <li><a href="#finding-analysis-summaries" id="toc-finding-analysis-summaries" class="nav-link" data-scroll-target="#finding-analysis-summaries">Finding analysis summaries</a></li>
  <li><a href="#using-the-catalogues" id="toc-using-the-catalogues" class="nav-link" data-scroll-target="#using-the-catalogues">Using the catalogues</a>
  <ul class="collapse">
  <li><a href="#mag-catalogues" id="toc-mag-catalogues" class="nav-link" data-scroll-target="#mag-catalogues">MAG Catalogues</a></li>
  <li><a href="#viral-catalogues" id="toc-viral-catalogues" class="nav-link" data-scroll-target="#viral-catalogues">Viral Catalogues</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/ebi-metagenomics/holofood-database/edit/main/docs/website.qmd" class="toc-action">Edit this page</a></p><p><a href="https://github.com/ebi-metagenomics/holofood-database/issues/new" class="toc-action">Report an issue</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Browsing the website</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="global-search" class="level2">
<h2 class="anchored" data-anchor-id="global-search">Global Search</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./img/website/global-search.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Screenshot of the global search box</figcaption>
</figure>
</div>
<p>There is a a site-wide (“global”) search box at the top of the data portal landing page and most other pages. This searches over <a href="./datasets.html#ena-samples">host and derived samples</a> accessions, titles, systems, and types; <a href="./datasets.html#metagenome-assembled-genome-mag-catalogues">genome</a> and <a href="./datasets.html#viral-fragment-catalogues">viral</a> catalogue ids, titles, biomes, and related data; MAG accessions and taxonomies; viral fragment ids and contig ids; and this documentation.</p>
<p>If you search for valid sample, animal, or MAG accessions, the search will take you straight to an appropriate view in the portal. Otherwise, you’ll see a normal list of search results.</p>
</section>
<section id="finding-samples" class="level2">
<h2 class="anchored" data-anchor-id="finding-samples">Finding Samples</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./img/website/find-samples.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Screenshot of sample listing webpage</figcaption>
</figure>
</div>
<p>Samples can be found from the “Chicken samples” and “Salmon samples” navigation items. On the samples listing pages, there are various filters available to limit which samples are shown in the table. For example, you can find all samples from a particular animal, or all <a href="./datasets.html#metabolights-metabolomics">metabolomics</a> type samples (each sample has a particular type).</p>
<p>Samples may be linked to metadata (almost always), metagenomic, metabolomic, or host-genomic datasets. Other sample types have their results data in the sample metadata.</p>
<section id="treatment-search" class="level3">
<h3 class="anchored" data-anchor-id="treatment-search">Treatment search</h3>
<p>Samples may also be search for by the host-animal’s treatment condition. The “Treatment search” filter is a case-insensitive text search over the samples’ animal treatment metadata. Specifically, it filters to samples whose host animal has all of the query strings in the animal’s BioSamples metadata fields: “Sampling time”, “Treatment name”, “Treatment description”, “Treatment code”, “Treatment concentration”.</p>
<p>This means you can search for e.g.&nbsp;<code>ferm algae 2.0% "day 60"</code> to find samples from animals treated with <strong>Ferm</strong>ented <strong>Algae</strong> at a <strong>2.0%</strong> concentration and collected on trial <strong>Day 60</strong>.</p>
<p>Note the use of lexical grouping around the “day 60” – this is useful when searching for, e.g.&nbsp;<code>day 0</code> which would otherwise partially match <code>day 60</code>.</p>
</section>
<section id="sample-detail" class="level3">
<h3 class="anchored" data-anchor-id="sample-detail">Sample detail</h3>
<p>To view detail about a particular sample, click “View” on its row of the table.</p>
<p>The sample detail page contains information help about the sample in the various supporting databases. Metadata (from BioSamples) is shown in full, in a table. Metagenomic and metabolomic data are shown in summary form, with links to the respective public websites where those analyses are held.</p>
<p>The <a href="./api.html">API</a> section shows the API endpoint for this particular sample. You can copy this into a script, for example, to programmatically pull the data.</p>
</section>
</section>
<section id="downloading-sample-lists-and-metadata" class="level2">
<h2 class="anchored" data-anchor-id="downloading-sample-lists-and-metadata">Downloading sample lists and metadata</h2>
<p>The complete sample list can be exported to TSV using the “Download all as TSV” button.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The “Download all as TSV” button does reflect any filters on the website table – it contains the complete list.</p>
</div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./img/website/sample-detail.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Screenshot of a sample detail page with TSV export option</figcaption>
</figure>
</div>
<p>The complete metadata for a sample can be downloaded using the “Download all as TSV” button within the “Sample metadata” section of a sample detail page. For sample types where the metadata also includes results data (like histology measurements), the section is instead called “Sample data”.</p>
</section>
<section id="finding-sample-data-in-other-public-repositories" class="level2">
<h2 class="anchored" data-anchor-id="finding-sample-data-in-other-public-repositories">Finding sample data in other public repositories</h2>
<section id="metagenomics" class="level3">
<h3 class="anchored" data-anchor-id="metagenomics">Metagenomics</h3>
<p>Some samples have <a href="./datasets.html#mgnify-metagenomics">metagenomics</a> data, in <a href="https://www.ebi.ac.uk/metagenomics">MGnify</a>. These can be found from the samples listing page by setting the <code>Sample type</code> filter to <code>metagenomic_assembly</code> or <code>metagenomic_amplicon</code>.</p>
<p>MGnify analyses of a sample (identified with <code>MGYA</code> accessions) are listed and linked to from the sample detail page.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./img/website/sample-with-metagenomics.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Screenshot of a metagenomic sample detail page</figcaption>
</figure>
</div>
</section>
<section id="metabolomics" class="level3">
<h3 class="anchored" data-anchor-id="metabolomics">Metabolomics</h3>
<p>Some other samples have <a href="./datasets.html#metabolights-metabolomics">metabolomics</a> data, in <a href="https://www.ebi.ac.uk/metabolights">MetaboLights</a>. These can be found form the samples listing page by setting the <code>Sample type</code> filter to <code>metabolomic</code>.</p>
<p>MetaboLights data are identified at a Project level, with <code>MTBLS</code> accessions. MetaboLights does not store samples as independent objects, instead it stores lists of samples and files (and more) for the project. So, the HoloFood data portal sample detail page shows a filtered table of the MetaboLights project’s files, that relate to this sample. Following these file links will download the file.</p>
<p>MetaboLights follows the <a href="https://isa-tools.org">ISA framework</a>, so the table shown is a collection of files from the one or more assay sheets relevant to this sample. Raw and derived files are available.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./img/website/sample-with-metabolomics.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Screenshot of a metabolomic sample detail page</figcaption>
</figure>
</div>
</section>
</section>
<section id="finding-samples-related-to-the-same-host-animal" class="level2">
<h2 class="anchored" data-anchor-id="finding-samples-related-to-the-same-host-animal">Finding samples related to the same host animal</h2>
<p>Each host animal (an individual chicken or salmon) is itself a BioSample. Host-level measurements of the animal and its tank/pen are available as metadata on the animal. The animal listing pages (accessed from links on the sample lists) and the animal detail pages each include information about the derived sample types available for a given animal.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./img/website/animal-detail.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Screenshot of an animal detail page</figcaption>
</figure>
</div>
</section>
<section id="finding-analysis-summaries" class="level2">
<h2 class="anchored" data-anchor-id="finding-analysis-summaries">Finding analysis summaries</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./img/website/sample-analysis-summaries.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Screenshot of a sample detail page with analysis documents links</figcaption>
</figure>
</div>
<p><a href="./datasets.html#summary-analyses">Analysis summaries</a> are linked to samples or catalogues. Any analysis summaries that mention a sample are shown at the bottom of the sample’s detail page.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./img/website/analysis-summary.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Screenshot of an analysis document</figcaption>
</figure>
</div>
<p>Analysis summaries also link back to the samples and/or catalogues they refer to. A complete list of analysis summaries can also be found from the navigation bar.</p>
</section>
<section id="using-the-catalogues" class="level2">
<h2 class="anchored" data-anchor-id="using-the-catalogues">Using the catalogues</h2>
<section id="mag-catalogues" class="level3">
<h3 class="anchored" data-anchor-id="mag-catalogues">MAG Catalogues</h3>
<p>Metagenome Assembled Genome (MAG) Catalogues are available for selected biomes. HoloFood MAGs are those created using only reads from HoloFood samples. However, there are other non-HoloFood public data available for the same biomes sampled by this project.</p>
<p>Each HoloFood MAG Catalogue therefore referenced a public MAG Catalogue in MGnify, which is a superset of the HoloFood data and other public data. This is linked from each catalogue page on the HoloFood Data Portal site.</p>
<p>Each MAG in the HoloFood catalogue references a MAG in the MGnify catalogue which represents the same species. In some cases, the HoloFood MAG is the best available sequence for that species level cluster, so the HoloFood MAG points to itself on the MGnify website. In other cases, a more complete, less contaminated, or isolate genome exists representing the same species, so the HoloFood MAG points to this better representative on MGnify.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./img/website/mag-catalogue.png" class="img-fluid figure-img"></p>
<figcaption class="figure-caption">Screenshot of a MAG catalogue</figcaption>
</figure>
</div>
<p>MAG Catalogues can be found from the “Genomes” navigation item, and then selecting a catalogue in the “Catalogues” sub-navigation. MAGs can be found by searching on accession or taxonomy, or for the accession of the cluster representative.</p>
<p>The MAGs in a catalogue can be downloaded as a TSV file, using the “Download all as TSV” button.</p>
</section>
<section id="viral-catalogues" class="level3">
<h3 class="anchored" data-anchor-id="viral-catalogues">Viral Catalogues</h3>
<p><img src="./img/website/viral-catalogue.png" class="img-fluid" alt="Screenshot of a viral catalogue"> Viral catalogues are lists of the unique (at species-level) viruses found in HoloFood samples. Viral catalogues can be found from the “Viruses” navigation item, and then selecting a catalogue in the “Catalogues” sub-navigation.</p>
<p>Viral fragment can be searched in various ways, like the parent contig ID, or the taxonomy of the host MAG. Host MAGs are only assigned where there is an exact match between the viral sequence and a host MAG.</p>
<p>By default, only species-level clusters are shown (a representative viral fragment is shown).</p>
<p>This can be changed either by clicking the “View cluster” link within the table, to see the remainder of that specific cluster, or by changing the “Cluster visibility” dropdown in the filters to the left of the table.</p>
<p>Pressing “View contig” on a viral fragment in the table opens the contig viewer. This loads the contig from MGnify, as well as MGnify’s annotations on the contig. These can be explored in more detail by clicking the parent contig in the table, which links to MGnify.</p>
<p>In the contig viewer, ViPhOG annotations are shown. These are unique to the data portal. The viral region is highlighted in green. The GFF containing the ViPhOG(s) can also be downloaded, by viewing a viral fragment and pressing “Download ViPhOGs GFF”.</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./datasets.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Datasets</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./api.html" class="pagination-link">
        <span class="nav-page-text">API for Programmatic Access</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">This project has received funding from the European Union’s Horizon 2020 research and innovation programme under grant agreement No 817729</div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>